python/lib/adaguc/CGIRunner.py

```diff
@@ -13,40 +14,48 @@
 import chardet
 from queue import Queue, Empty  # python 3.x
 import re
-#from numba import jit
 
+# from numba import jit
+
+ADAGUC_NUMPARALLELPROCESSES = os.getenv("ADAGUC_NUMPARALLELPROCESSES", "4")
+# print("ADAGUC_NUMPARALLELPROCESSES", ADAGUC_NUMPARALLELPROCESSES)
+sem = asyncio.Semaphore(int(ADAGUC_NUMPARALLELPROCESSES))
 
-class CGIRunner:
 
+class CGIRunner:
     """
-      Run the CGI script with specified URL and environment. Stdout is captured and put in a BytesIO object provided in output
+    Run the CGI script with specified URL and environment. Stdout is captured and put in a BytesIO object provided in output
     """
 
-    def run(self, cmds, url, output, env=[],  path=None, isCGI=True, timeout=300):
+    async def run(self, cmds, url, output, env=[], path=None, isCGI=True, timeout=300):
         localenv = {}
         if url != None:
-            localenv['QUERY_STRING'] = url
+            localenv["QUERY_STRING"] = url
         else:
-            localenv['QUERY_STRING'] = ""
+            localenv["QUERY_STRING"] = ""
         if path != None:
-            localenv['SCRIPT_NAME'] = "/myscriptname"
+            localenv["SCRIPT_NAME"] = "/myscriptname"
             # SCRIPT_NAME [/cgi-bin/autoresource.cgi], REQUEST_URI [/cgi-bin/autoresource.cgi/opendap/clipc/combinetest/wcs_nc2.nc.das]
-            localenv['REQUEST_URI'] = "/myscriptname/" + path
+            localenv["REQUEST_URI"] = "/myscriptname/" + path
         localenv.update(env)
 
         # Execute adaguc-server binary
-        ON_POSIX = 'posix' in sys.builtin_module_names
-        process = Popen(cmds, stdout=PIPE, stderr=PIPE, env=localenv, close_fds=ON_POSIX)
-        
-        try:
-            (processOutput, processError) = process.communicate(timeout=timeout)
-        except TimeoutExpired:
-            process.kill()
-            process.communicate()
-            output.write(b'TimeOut')
-            return 1, [], None
+        ON_POSIX = "posix" in sys.builtin_module_names
+        async with sem:
+            process = await asyncio.create_subprocess_exec(
+                *cmds, stdout=PIPE, stderr=PIPE, env=localenv, close_fds=ON_POSIX
+            )
+
+            try:
+                asyncio.wait_for(process, timeout=timeout)
```

Comment:
It seems that that the timeout is not working. When I place an infinite loop in adguc-server, the process will hang.